cp_diffdock_api.py

```diff
@@ -0,0 +1,261 @@
+    A protein or ligand part of a docking job, including the file path.
+    """
+    name: str
+    label: str = ""
+    path: str = ""
+    error: str = ""
+
+    def ok(self):
+        return not self.error
+
+    def __iter__(self):
+        # Specific about fields here so error isn't included when unpacking
+        return iter([self.name, self.label, self.path])
+
+
+@dataclass
+class PreppedComplex:
+    """
+    A protein-ligand complex to be docked, including the protein and ligand parts.
+    This contains the data for the DiffDock csv input file.
+    """
+    name: str
+    protein: PreppedComplexUnit
+    ligand: PreppedComplexUnit
+
+    def __iter__(self):
+        # Specific about fields here to avoid converting protein and ligand to dict
+        return iter([self.name, self.protein, self.ligand])
+
+
+@dataclass
+class PreppedRequest:
+    """
+    A DiffDock request after preparing the pdb, sdf, and csv input files.
+    """
+    csv_file: str
+    out_dir: str
+    complex_infos: List[PreppedComplex]
+
+    def __iter__(self):
+        # Specific about fields here to avoid converting complex_infos to dict
+        return iter([self.csv_file, self.out_dir, self.complex_infos])
+
+@dataclass
+class DiffDockOptions:
+    work_dir: Path
+
+    @staticmethod
+    def make(baseDir=None, requestId=None):
+        # get_id needs to run every time, so don't use default arguments
+        if baseDir is None:
+            baseDir = f"/tmp/diffdock{os.getpid()}"
+        if requestId is None:
+            requestId = requestIdGenerator.get_id()
+        work_dir = Path(baseDir) / f"dock_{requestId}"
+        return DiffDockOptions(
+            work_dir=work_dir
+        )
+
+
+class DiffDockNothingToDoException(Exception):
+    pass
+
+class DiffDockSetupException(Exception):
+    pass
+
+
+class DiffDockApi:
+    """
+    Wrapper for running DiffDock with DiffDockProtocol objects for requests and responses.
+    """
+    @staticmethod
+    def run_diffdock(request: DiffDockProtocol.Request, options: DiffDockOptions=None) -> DiffDockProtocol.Response:
+        try:
+            log.info("diffdock preparing request...")
+            preppedRequest = DiffDockApi.prepare_request(request, options)
+
+            args = InferenceArgs({
+                "protein_ligand_csv": preppedRequest.csv_file,
+                "out_dir": preppedRequest.out_dir,
+                "add_hs": request.add_hs,
+                "keep_hs": request.keep_hs,
+                "keep_src_3d": request.keep_src_3d,
+                "samples_per_complex": request.samples_per_complex,
+            })
+            log.info("diffdock running inference...")
+            RunInference(args)
+            response = DiffDockApi.process_results(preppedRequest)
+            log.info("diffdock prepared response.")
+            return response
+        except DiffDockNothingToDoException:
+            return DiffDockProtocol.Response.makeResults()
+        except Exception as ex:
```

Comment:
Maybe this wants to log the exception stack trace with python's traceback.format_exc() or something like that. The docs for traceback are hard for me to find what I want.